### DISCUSSION TITLE
SSLCertVerificationError can be handled by RequestError

### DISCUSSION THREAD

**Discussion by nkhitrov** [encode/httpx#2682] [Potential Issue]
I use the `AsyncClient.request()` method to call one of the internal services. Today due to Kubernetes configuration error I got `SSLCertVerificationError`.

I was very surprised because there is a `RequestError` exception handling in the code. The description of this error says that it can catch all possible errors:
```
Base class for all exceptions that may occur when issuing a `.request()`
```

Why this ssl error is not handled in the same way as connection errors? 
Maybe this behaviour should be changed?

**Comment by lovelydinosaur** [comment]
> Why this ssl error is not handled in the same way as connection errors?

Where would you look first to answer this question?

What other information do you have available that might help you resolve this?

**Reply by nkhitrov** [reply]
I think we should add a class for this ssl error and inherit it from the [NetworkError](https://github.com/encode/httpx/blob/master/httpx/_exceptions.py#L12)

**Reply by lovelydinosaur** [reply]
What is the traceback that's raised?

**Reply by nkhitrov** [reply]
```python
 File "app/core/advertising/adapters.py", line 62, in _request
    response = await self.http_client.request(method=method, url=url, params=params, headers=headers)
  File "httpx/_client.py", line 1533, in request
    return await self.send(request, auth=auth, follow_redirects=follow_redirects)
  File "httpx/_client.py", line 1620, in send
    response = await self._send_handling_auth(
  File "httpx/_client.py", line 1648, in _send_handling_auth
    response = await self._send_handling_redirects(
  File "httpx/_client.py", line 1685, in _send_handling_redirects
    response = await self._send_single_request(request)
  File "httpx/_client.py", line 1722, in _send_single_request
    response = await transport.handle_async_request(request)
  File "/app/.venv/lib/python3.11/site-packages/opentelemetry/instrumentation/httpx/__init__.py", line 431, in handle_async_request
    response = await self._transport.handle_async_request(
  File "httpx/_transports/default.py", line 353, in handle_async_request
    resp = await self._pool.handle_async_request(req)
  File "httpcore/_async/connection_pool.py", line 253, in handle_async_request
    raise exc
  File "httpcore/_async/connection_pool.py", line 237, in handle_async_request
    response = await connection.handle_async_request(request)
  File "httpcore/_async/connection.py", line 86, in handle_async_request
    raise exc
  File "httpcore/_async/connection.py", line 63, in handle_async_request
    stream = await self._connect(request)
  File "httpcore/_async/connection.py", line 150, in _connect
    stream = await stream.start_tls(**kwargs)
  File "httpcore/backends/asyncio.py", line 78, in start_tls
    raise exc
  File "httpcore/backends/asyncio.py", line 69, in start_tls
    ssl_stream = await anyio.streams.tls.TLSStream.wrap(
  File "anyio/streams/tls.py", line 122, in wrap
    await wrapper._call_sslobject_method(ssl_object.do_handshake)
  File "anyio/streams/tls.py", line 130, in _call_sslobject_method
    result = func(*args)
  File "ssl.py", line 979, in do_handshake
    self._sslobj.do_handshake()
```

**Reply by nkhitrov** [reply]
we also use opentelemetry httpx instrumentation, but it should not affect the request

**Reply by nkhitrov** [reply]
@tomchristie hi! What do you think about this issue? Can I make a PR with fix?

**Reply by lovelydinosaur** [reply]
> Can I make a PR with fix?

Yes. We could choose to use `ConnectError` for this, so long as it includes proper error details.

**Reply by nkhitrov** [reply]
I read `httpx` and `httpcore` code and I think we should add `OSError` to exception mapping  [here](https://github.com/encode/httpcore/blob/b1137996b1828988a584daa258e9d7afefd2e2d7/httpcore/backends/asyncio.py#L62-L79) like [in sync backend](https://github.com/encode/httpcore/blob/b1137996b1828988a584daa258e9d7afefd2e2d7/httpcore/backends/sync.py#L52).

**Reply by lovelydinosaur** [reply]
That looks correct to me, yep.

**Comment by lahwaacz** [comment]
This is a problem also because the behavior is different for `AsyncClient` (gives `ssl.SSLError` or its subclass) and `Client` (gives `httpx.ConnectError`). Consider these repro cases:

1. sync version:
```python
import httpx
import ssl
ssl_context = httpx.create_ssl_context(ssl.PROTOCOL_TLS)
ssl_context.options |= ssl.OP_NO_TLSv1 | ssl.OP_NO_TLSv1_1 | ssl.OP_NO_TLSv1_2 | ssl.OP_NO_TLSv1_3
client = httpx.Client(verify=ssl_context)
client.get("https://www.howsmyssl.com/a/check")
```
gives `httpx.ConnectError: [SSL: NO_PROTOCOLS_AVAILABLE] no protocols available (_ssl.c:997)`

2. async version:
```python
import asyncio
import httpx
import ssl
ssl_context = httpx.create_ssl_context(ssl.PROTOCOL_TLS)
ssl_context.options |= ssl.OP_NO_TLSv1 | ssl.OP_NO_TLSv1_1 | ssl.OP_NO_TLSv1_2 | ssl.OP_NO_TLSv1_3
client = httpx.AsyncClient(verify=ssl_context)
async def main():
    await client.get("https://www.howsmyssl.com/a/check")
loop = asyncio.get_event_loop()
loop.run_until_complete(main())
```
gives `ssl.SSLError: [SSL: NO_PROTOCOLS_AVAILABLE] no protocols available (_ssl.c:997)`
